### ISSUE TITLE
Request for adding training scripts for SPIN and SPPO

### ISSUE THREAD

**Issue by jkx19** [huggingface/trl#2958]
### Feature request

I am one of the authors of paper SPIN (https://arxiv.org/abs/2401.01335) and SPPO (https://arxiv.org/abs/2405.00675). I am delighted to see that the training loss for a single iteration of SPIN and SPPO has been implemented in the DPO trainer. However, since both SPIN and SPPO are iterative algorithms, would it be possible to include training scripts that reflect the full pipeline? I appreciate your consideration!

### Motivation

Currently, to run SPIN and SPPO on trl, I need to manually setup the configuration and write a scripts to iteratively run the dpo training scripts. For example, to run SPPO for 3 iteration, I might need a script like the follows
```
iter_num=3
for i in $(seq 1 $iter_num); do
    if [ "$i" -eq 1 ]; then
        MODEL="{{base_model}}"
    else
        MODEL=$OUTPUT_DIR
    fi
    OUTPUT_DIR="checkpoints/SPPO-Iter${i}"
    OUT="data-llama-3-8b-instruct-sppo-iter${i}"

    bash generate.sh --model $MODEL --out_path $OUT # Generate the responses
    python trl/scripts/dpo.py --loss_type sppo_hard # Run DPO trainer, Other arguments omitted
done
```

Would it be possible to include a script that outlines the full training process of SPIN/SPPO? This would be greatly helpful for users, as it would eliminate the need to manually write such a script.

### Your contribution

I can help prepare the PR if the feature request is approved.

**Comment by albertvillanova** [comment]
Hey @jkx19, sorry for the very slow reply here — we're catching up on the backlog. 

Thanks for this, and for actually taking a shot at it in #3011. The per-iteration SPIN/SPPO objective is already in TRL (the `sppo_hard` loss, and SPIN via the DPO loss), and the SPPO reproduction config is documented in the paper index: https://github.com/huggingface/trl/blob/44596766101a123f1a3ff70b0aefa54d926179ad/docs/source/paper_index.md#L1033-L1050. We've made a deliberate call to keep the *outer* iterative loop (regenerate responses -> rescore -> retrain) out of core — it's inherently pipeline- and infra-specific (generation backend, PairRM scoring, dataset regeneration), and as a bundled example script it tends to bit-rot fast, which is why #3011 stalled. 

So I'll close this as won't-implement for the core repo. Really appreciate the contribution!
